x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -179,8 +180,10 @@ public void execute(EsqlQueryRequest request, EsqlExecutionInfo executionInfo, P
         analyzedPlan(parsed, executionInfo, request.filter(), new EsqlCCSUtils.CssPartialErrorsActionListener(executionInfo, listener) {
             @Override
             public void onResponse(LogicalPlan analyzedPlan) {
+                assert ThreadPool.assertCurrentThreadPool(ThreadPool.Names.SEARCH);
                 SubscribableListener.<LogicalPlan>newForked(l -> preOptimizedPlan(analyzedPlan, l))
-                    .<LogicalPlan>andThen((l, p) -> preMapper.preMapper(optimizedPlan(p), l))
+                    .andThenApply(p -> optimizedPlan(p))
+                    .<LogicalPlan>andThen(planExecutor, null, (l, p) -> preMapper.preMapper(p, l))
```

Comment:
Should we specify the thread pool in the next step instead because `preMapper.preMapper()` can switch the executing thread during rewrite. https://github.com/elastic/elasticsearch/blob/4790cf4c3edc1ee7ee30bc15ba482e0064d36db9/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryBuilderResolver.java#L44-L48

Reply:
This is also a place that allow to narrow down the list of thread everywhere else in this change.